### PR TITLE
engine: add timestamp to user created snapshot

### PR DESCRIPTION
**Pull request by DamiaSan**
#### Which issue(s) this PR fixes:

Issue #7641

#### Special notes for your reviewer:

I added the timestamp only to the snapshots created by the user, because they are the only ones listed in the UI.

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -445,11 +456,16 @@ func (c *SPDKClient) EngineSnapshotCreate(name, snapshotName string, opts *api.S
...
 
-	resp, err := client.EngineSnapshotCreate(ctx, &spdkrpc.SnapshotRequest{
+	snapshotRequest := spdkrpc.SnapshotRequest{
 		Name:         name,
 		SnapshotName: snapshotName,
 		UserCreated:  opts != nil && opts.UserCreated,
-	})
+	}
+	if opts != nil {
+		snapshotRequest.SnaphotTimestamp = opts.Timestamp
```
Comment (derekbit) [inline pkg/client/client.go:465]: Same as above

File: pkg/spdk/server.go [hunk 2]
```diff
@@ -452,7 +452,12 @@ func (s *Server) ReplicaSnapshotCreate(ctx context.Context, req *spdkrpc.Snapsho
 		return nil, grpcstatus.Errorf(grpccodes.NotFound, "cannot find replica %s during snapshot create", req.Name)
 	}
 
-	return r.SnapshotCreate(spdkClient, req.SnapshotName, &api.SnapshotOptions{UserCreated: req.UserCreated})
+	opts := api.SnapshotOptions{
+		UserCreated: req.UserCreated,
+		Timestamp:   req.SnaphotTimestamp,
+	}
+
+	return r.SnapshotCreate(spdkClient, req.SnapshotName, &opts)
```
Comment (derekbit) [inline pkg/spdk/server.go:460]: Why not
```suggestion
	opts := &api.SnapshotOptions{
		UserCreated: req.UserCreated,
		Timestamp:   req.SnaphotTimestamp,
	}

	return r.SnapshotCreate(spdkClient, req.SnapshotName, opts)
```

File: pkg/spdk/server.go [hunk 3]
```diff
@@ -660,7 +665,12 @@ func (s *Server) ReplicaRebuildingDstSnapshotCreate(ctx context.Context, req *sp
 		return nil, grpcstatus.Errorf(grpccodes.NotFound, "cannot find replica %s during rebuilding dst snapshot create", req.Name)
 	}
 
-	if err = r.RebuildingDstSnapshotCreate(spdkClient, req.SnapshotName, &api.SnapshotOptions{UserCreated: req.UserCreated}); err != nil {
+	opts := api.SnapshotOptions{
+		UserCreated: req.UserCreated,
+		Timestamp:   req.SnaphotTimestamp,
+	}
+
+	if err = r.RebuildingDstSnapshotCreate(spdkClient, req.SnapshotName, &opts); err != nil {
```
Comment (derekbit) [inline pkg/spdk/server.go:673]: Same as above.

File: pkg/client/client.go [hunk 4]
```diff
@@ -147,11 +147,17 @@ func (c *SPDKClient) ReplicaSnapshotCreate(name, snapshotName string, opts *api.
 	ctx, cancel := context.WithTimeout(context.Background(), GRPCServiceTimeout)
 	defer cancel()
 
-	_, err := client.ReplicaSnapshotCreate(ctx, &spdkrpc.SnapshotRequest{
+	snapshotRequest := spdkrpc.SnapshotRequest{
 		Name:         name,
 		SnapshotName: snapshotName,
 		UserCreated:  opts != nil && opts.UserCreated,
```
[the remark this answers is not in the document]
Comment (DamiaSan) [reply pkg/client/client.go]: Good improvement, thanks

File: pkg/client/client.go (hunk 1 above)
Comment (DamiaSan) [reply pkg/client/client.go]: Here `SnapshotTimestamp` is not needed, I removed it. Actually, not even `UserCreated` is needed, so the `opts` parameter is unnecessary, maybe it is better to remove it after this PR will be merged. WDYT?

File: pkg/spdk/engine.go [hunk 5]
```diff
@@ -772,7 +772,10 @@ func (e *Engine) ReplicaAddStart(spdkClient *spdkclient.Client, replicaName, rep
 
 	// TODO: For online rebuilding, the IO should be paused first
 	snapshotName := GenerateRebuildingSnapshotName()
-	updateRequired, err = e.snapshotOperationWithoutLock(spdkClient, replicaClients, snapshotName, SnapshotOperationCreate, nil)
+	opts := api.SnapshotOptions{
+		Timestamp: time.Now().String(),
+	}
+	updateRequired, err = e.snapshotOperationWithoutLock(spdkClient, replicaClients, snapshotName, SnapshotOperationCreate, &opts)
```
Comment (derekbit) [inline pkg/spdk/engine.go:778]: nit:
```suggestion
	opts := &api.SnapshotOptions{
		Timestamp: time.Now().String(),
	}
	updateRequired, err = e.snapshotOperationWithoutLock(spdkClient, replicaClients, snapshotName, SnapshotOperationCreate, opts)
```

File: pkg/client/client.go (hunk 1 above)
Comment (derekbit) [reply pkg/client/client.go]: > not even UserCreated is needed

Can you elaborate more on it?

Comment (DamiaSan) [reply pkg/client/client.go]: Both user created and snapshot timestamp flags are internally resolved in `longhorn-spdk-engine`, so it is not needed to insert them as parameters in the function `EngineSnapshotCreate`, which is called by `longhorn-instance-manager`.

Comment (derekbit) [reply pkg/client/client.go]: I see. Then, we can remove the parameter `opts`.

Comment (derekbit) [reply pkg/client/client.go]: Why not remove it in this PR?

**Review by derekbit — APPROVED** [review]
LGTM

**Pull request merged**